### PR TITLE
JEP 451 skips the warning message if the agent library was loaded

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Updated `isAgentLibraryLoaded()` to test if an agent library was loaded via invoking `j9sl_open_shared_library()` with `J9PORT_SLOPEN_NO_LOAD` flag;
Updated to print the warning message only when `loadAgentLibraryOnAttach()` was successful.

Passed [JDK21 DynamicLoadWarningTest](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/openj9/test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java) at `AIX/Linux/OSX/Windows` platforms.

Note: `AIX` is excluded from [the test](https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/c0d4f334bfe09ed00cd9b26e2cc41c0ba227d484/test/jdk/com/sun/tools/attach/warnings/DynamicLoadWarningTest.java#L127-L128) if an agent library was previously loaded.

depends on https://github.com/eclipse/omr/pull/7093

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
Setting as draft until the OMR change promotes.

File: runtime/jvmti/jvmtiStartup.c [hunk 1]
```diff
@@ -1062,10 +1061,57 @@ createXrunLibraries(J9JavaVM * vm)
 	return JNI_OK;
 }
 
+/**
+ * Test if an agent library was already loaded.
+ *
+ * @param vm Java VM
+ * @param library library name
+ * @param decorate the flag indicating if change the library path/name
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiStartup.c]: Changed `decorate` to `BOOLEAN`, and updated the comment.

File: runtime/jvmti/jvmtiStartup.c [hunk 2]
```diff
@@ -1062,10 +1061,57 @@ createXrunLibraries(J9JavaVM * vm)
...
+	if (NULL == findAgentLibrary(vm, library, strlen(library))) {
+		/* J9PORT_SLOPEN_NO_LOAD is only supported on Linux/OSX/Win32 platforms. */
+#if defined(LINUX) || defined(OSX) || defined(WIN32)
+		int i = 0;
+		const char *systemAgentName = NULL;
+		char *fullLibName = NULL;
+		UDATA handle = 0;
+		/* One of RTLD_LAZY and RTLD_NOW must be included in the flag. */
+		UDATA openFlags = J9PORT_SLOPEN_NO_LOAD | OMRPORT_SLOPEN_LAZY;
+		char *agentPath = NULL;
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiStartup.c]: Removed `fullLibName`.

File: runtime/jvmti/jvmtiStartup.c [hunk 3]
```diff
@@ -1062,10 +1061,57 @@ createXrunLibraries(J9JavaVM * vm)
...
+			openFlags |= J9PORT_SLOPEN_DECORATE;
+		}
+		while (NULL != (systemAgentName = systemAgentNames[i++])) {
+			if (0 == strcmp(library, systemAgentName)) {
+				fullLibName = prependSystemAgentPath(vm, library);
+				break;
+			}
+		}
+		agentPath = (NULL != fullLibName) ? fullLibName : (char*)library;
+		if (0 == j9sl_open_shared_library(agentPath, &handle, openFlags)) {
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiStartup.c:1098]: Added `j9sl_close_shared_library()`.

File: runtime/jvmti/jvmtiStartup.c [hunk 4]
```diff
@@ -1062,10 +1061,56 @@ createXrunLibraries(J9JavaVM * vm)
...
+ * @return TRUE if the agent library was loaded successfully, FALSE otherwise
+ */
 static BOOLEAN
-isAgentLibraryLoaded(J9JavaVM *vm, const char *library)
+isAgentLibraryLoaded(J9JavaVM *vm, const char *library, BOOLEAN decorate)
 {
-	return NULL != findAgentLibrary(vm, library, strlen(library));
+	BOOLEAN result = FALSE;
+
+	if (NULL == findAgentLibrary(vm, library, strlen(library))) {
```
Comment (keithc-ca) [inline runtime/jvmti/jvmtiStartup.c:1078]: Suggest inverting this so the condition is close to the beginning of both branches:
```c
	if (NULL != findAgentLibrary(vm, library, strlen(library))) {
		result = TRUE;
	} else {
		/* J9PORT_SLOPEN_NO_LOAD is only supported on Linux/OSX/Win32 platforms. */
```

File: runtime/jvmti/jvmtiStartup.c [hunk 5]
```diff
@@ -1062,10 +1061,56 @@ createXrunLibraries(J9JavaVM * vm)
...
+		/* J9PORT_SLOPEN_NO_LOAD is only supported on Linux/OSX/Win32 platforms. */
+#if defined(LINUX) || defined(OSX) || defined(WIN32)
+		int i = 0;
```
Comment (keithc-ca) [inline runtime/jvmti/jvmtiStartup.c:1081]: Suggest `UDATA` instead of `int`.

File: runtime/jvmti/jvmtiStartup.c [hunk 6]
```diff
@@ -1062,10 +1061,56 @@ createXrunLibraries(J9JavaVM * vm)
...
+
+		if (decorate) {
+			openFlags |= J9PORT_SLOPEN_DECORATE;
+		}
+		while (NULL != (systemAgentName = systemAgentNames[i++])) {
+			if (0 == strcmp(library, systemAgentName)) {
+				agentPath = prependSystemAgentPath(vm, library);
+				break;
+			}
+		}
```
Comment (keithc-ca) [inline runtime/jvmti/jvmtiStartup.c:1097]: Assignments, increments and tests should all be separate:
```c
		for (i = 0;; ++i) {
			const char *systemAgentName = systemAgentNames[i];
			if (NULL == systemAgentName) {
				break;
			}
			if (0 == strcmp(library, systemAgentName)) {
				agentPath = prependSystemAgentPath(vm, library);
				break;
			}
		}
```

File: runtime/vm/BytecodeInterpreter.hpp [hunk 7]
```diff
@@ -4873,20 +4873,26 @@ class INTERPRETER_CLASS
...
 					J9JavaVM *vm = _vm;
 					/* No warning if -XX:+EnableDynamicAgentLoading is specified. */
 					if (0 > FIND_ARG_IN_VMARGS(EXACT_MATCH, VMOPT_XXENABLEDYNAMICAGENTLOADING, NULL)) {
 						/* No warning if the same agent is already loaded. */
-						if (!vm->isAgentLibraryLoaded(vm, agentLibraryUTF)) {
-							fprintf(stderr, "WARNING: A JVM TI agent has been loaded dynamically (%s)\n", agentOptionsUTF);
-							fprintf(stderr, "WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning\n");
-							fprintf(stderr, "WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information\n");
-							fprintf(stderr, "WARNING: Dynamic loading of agents will be disallowed by default in a future release\n");
+						if (!vm->isAgentLibraryLoaded(vm, agentLibraryUTF, 0 != decorate)) {
```
Comment (keithc-ca) [inline runtime/vm/BytecodeInterpreter.hpp:4882]: `0` is not an appropriate value to compare with something of type `jboolean`,
the final argument should be `JNI_FALSE != decorate`.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/jvmti/jvmtiStartup.c (hunk 4 above)
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiStartup.c]: Inverted

File: runtime/jvmti/jvmtiStartup.c (hunk 5 above)
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiStartup.c]: Updated

File: runtime/jvmti/jvmtiStartup.c (hunk 6 above)
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiStartup.c]: Updated

File: runtime/vm/BytecodeInterpreter.hpp (hunk 7 above)
Comment (JasonFengJ9) [reply runtime/vm/BytecodeInterpreter.hpp]: Updated.

**Comment by pshipton** [comment]
The OMR change is promoted.

**Review by pshipton — APPROVED** [review]

**Comment by keithc-ca** [comment]
Jenkins test sanity.openjdk alinux64 jdk21

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**